# Common table expressions

Using the northwind database

In [1]:
import pymysql
import sqlalchemy as alch # python -m pip install --upgrade 'sqlalchemy<2.0'
from getpass import getpass
import pandas as pd

## 1. Write a CTE that lists the names and quantities of products with a unit price greater than $50.

In [2]:
password = getpass()

In [5]:
dbName = "northwind"
connectionData=f"mysql+pymysql://root:{password}@localhost/{dbName}"
engine = alch.create_engine(connectionData)
query = "SELECT * FROM employees;"
pd.read_sql_query(query, engine)

EmployeeID   LastName FirstName  BirthDate        Photo  \
0           1    Davolio     Nancy 1968-12-08   EmpID1.pic   
1           2     Fuller    Andrew 1952-02-19   EmpID2.pic   
2           3  Leverling     Janet 1963-08-30   EmpID3.pic   
3           4    Peacock  Margaret 1958-09-19   EmpID4.pic   
4           5   Buchanan    Steven 1955-03-04   EmpID5.pic   
5           6     Suyama   Michael 1963-07-02   EmpID6.pic   
6           7       King    Robert 1960-05-29   EmpID7.pic   
7           8   Callahan     Laura 1958-01-09   EmpID8.pic   
8           9  Dodsworth      Anne 1969-07-02   EmpID9.pic   
9          10       West      Adam 1928-09-19  EmpID10.pic   

                                               Notes  
0  Education includes a BA in psychology from Col...  
1  Andrew received his BTS commercial and a Ph.D....  
2  Janet has a BS degree in chemistry from Boston...  
3  Margaret holds a BA in English literature from...  
4  Steven Buchanan graduated from St. Andrews Uni...  
5  Michael is a graduate of Sussex University (MA...  
6  Robert King served in the Peace Corps and trav...  
7  Laura received a BA in psychology from the Uni...  
8  Anne has a BA degree in English from St. Lawre...  
9                                       An old chum.

In [13]:
query = """
SELECT productname, unit
FROM products
WHERE products.price > 50
"""
df = pd.read_sql_query(query, engine)
df

productname                  unit
0          Mishi Kobe Niku      18 - 500 g pkgs.
1         Carnarvon Tigers            16 kg pkg.
2   Sir Rodney's Marmalade         30 gift boxes
3  Thüringer Rostbratwurst  50 bags x 30 sausgs.
4            Côte de Blaye    12 - 75 cl bottles
5    Manjimup Dried Apples      50 - 300 g pkgs.
6     Raclette Courdavault             5 kg pkg.

#### Expected results

|ProductName                     |Unit|
|--------------------------------|----|
|Mishi Kobe Niku                 |18 - 500 g pkgs.|
|Carnarvon Tigers                |16 kg pkg.|
|Sir Rodney's Marmalade          |30 gift boxes|
|Thüringer Rostbratwurst         |50 bags x 30 sausgs.|
|Côte de Blaye                   |12 - 75 cl bottles|
|Manjimup Dried Apples           |50 - 300 g pkgs.|
|Raclette Courdavault            |5 kg pkg.|


## 2. What are the top 5 most profitable products?

In [37]:
query = """
SELECT DISTINCT(products.ProductID), products.ProductName, (products.unit * products.price) AS TotalRevenue
FROM products 
JOIN orderdetails ON products.productid = orderdetails.productid
ORDER BY totalrevenue DESC
LIMIT 5
"""

df = pd.read_sql_query(query, engine)
df


ProductID              ProductName  TotalRevenue
0         39         Chartreuse verte       13500.0
1         76             Lakkalikööri        9000.0
2         29  Thüringer Rostbratwurst        6200.0
3         27       Schoggi Schokolade        4400.0
4         38            Côte de Blaye        3168.0

#### Expected results
Alice Mutton is 6th

|ProductID                       |ProductName|TotalRevenue|
|--------------------------------|-----------|------------|
|-                             |-|-       |
|-                           |-|-       |
|-                            |-|-       |
|-                             |-|-       |
|-                              |-|-       |
|17                              |Alice Mutton|12909       |

## 3. Write a CTE that lists the top 5 categories by the number of products they have.

In [35]:
query =  """
SELECT categories.categoryname, COUNT(products.productname) AS count_productname
FROM  categories
JOIN products ON categories.categoryid = products.categoryid
GROUP BY categories.categoryname 
ORDER BY count_productname DESC
LIMIT 5
"""

df = pd.read_sql_query(query, engine)
df


categoryname  count_productname
0     Confections                 13
1       Beverages                 12
2      Condiments                 12
3         Seafood                 12
4  Dairy Products                 10

#### Expected result
|CategoryName                    |ProductCount|
|--------------------------------|------------|
|Confections                     |13          |
|Beverages                       |12          |
|Condiments                      |12          |
|Seafood                         |12          |
|Dairy Products                  |10          |


##  4. Write a CTE that shows the average order quantity for each product category.

In [39]:
query =  """
SELECT categories.CategoryName, AVG(orderdetails.Quantity)
FROM categories
JOIN products ON categories.categoryID = products.categoryID
JOIN orderdetails ON products.productID = orderdetails.productID
GROUP BY categoryname
"""

df = pd.read_sql_query(query, engine)
df


CategoryName  AVG(orderdetails.Quantity)
0       Beverages                     24.6129
1      Condiments                     28.2245
2     Confections                     25.1190
3  Dairy Products                     26.0100
4  Grains/Cereals                     21.7143
5    Meat/Poultry                     25.7600
6         Produce                     21.6667
7         Seafood                     21.5672

|CategoryName                    |AvgOrderQuantity|
|--------------------------------|----------------|
|Beverages                       |24.6129         |
|Condiments                      |28.2245         |
|Confections                     |25.1190         |
|Dairy Products                  |26.0100         |
|Grains/Cereals                  |21.7143         |
|Meat/Poultry                    |25.7600         |
|Produce                         |21.6667         |
|Seafood                         |21.5672         |


# 5. Create a CTE to calculate the average order amount for each customer.

In [44]:
query =  """
SELECT customers.CustomerID, customers.CustomerName, AVG(orderdetails.orderID) AS AvgOrderAmount
FROM customers
JOIN orders ON customers.customerID = orders.customerID
JOIN orderdetails ON orders.orderID = orderdetails.orderID
GROUP BY CustomerID
ORDER BY AvgOrderAmount DESC;
"""

df = pd.read_sql_query(query, engine)
df

CustomerID                CustomerName  AvgOrderAmount
0           79          Toms Spezialitäten      10438.0000
1           16       Consolidated Holdings      10435.0000
2           31         Gourmet Lanchonetes      10423.0000
3           27              Franchi S.p.A.      10422.0000
4           54      Océano Atlántico Ltda.      10409.0000
..         ...                         ...             ...
69          33        GROSELLA-Restaurante      10268.0000
70          81      Tradição Hipermercados      10263.3333
71          13  Centro comercial Moctezuma      10259.0000
72          34               Hanari Carnes      10251.5000
73          90                 Wilman Kala      10248.0000

[74 rows x 3 columns]

|CustomerID                      |CustomerName|AvgOrderAmount|
|--------------------------------|------------|--------------|
|59                              |Piccolo und mehr|4014.2500     |
|73                              |Simons bistro|2444.3333     |
|62                              |Queen Cozinha|1991.6667     |
|51                              |Mère Paillarde|1673.8571     |
|71                              |Save-a-lot Markets|1407.2500     |
|76                              |Suprêmes délices|1345.8333     |
|81                              |Tradição Hipermercados|1315.6667     |
|7                               |Blondel père et fils|1174.4615     |
|89                              |White Clover Markets|1112.5000     |
|55                              |Old World Delicatessen|1079.5000     |
|20                              |Ernst Handel|1018.0000     |
|19                              |Eastern Connection|1004.8000     |
|68                              |Richter Supermarkt|976.6667      |
|72                              |Seven Seas Imports|934.0000      |
|63                              |QUICK-Stop  |908.5500      |
|25                              |Frankenversand|895.0667      |
|75                              |Split Rail Beer & Ale|854.7692      |
|65                              |Rattlesnake Canyon Grocery|838.2727      |
|52                              |Morgenstern Gesundkost|754.0000      |
|9                               |Bon app''   |750.4286      |
|37                              |Hungry Owl All-Night Grocers|733.5714      |
|35                              |HILARIÓN-Abastos|722.8333      |
|33                              |GROSELLA-Restaurante|690.0000      |
|34                              |Hanari Carnes|681.0000      |
|15                              |Comércio Mineiro|677.5000      |
|23                              |Folies gourmandes|672.3333      |
|10                              |Bottom-Dollar Marketse|664.0000      |
|31                              |Gourmet Lanchonetes|641.0000      |
|60                              |Princesa Isabel Vinhoss|628.5000      |
|5                               |Berglunds snabbköp|601.5556      |
|46                              |LILA-Supermercado|564.7692      |
|44                              |Lehmanns Marktstand|545.7500      |
|30                              |Godos Cocina Típica|515.0000      |
|3                               |Antonio Moreno Taquería|504.0000      |
|88                              |Wellington Importadora|503.8333      |
|47                              |LINO-Delicateses|500.0000      |
|56                              |Ottilies Käseladen|500.0000      |
|87                              |Wartian Herkku|492.5833      |
|24                              |Folk och fä HB|480.2222      |
|49                              |Magazzini Alimentari Riuniti|464.2857      |
|86                              |Die Wandernde Kuh|441.9091      |
|41                              |La maison d''Asie|437.8182      |
|8                               |Bólido Comidas preparadas|416.0000      |
|83                              |Vaffeljernet|411.0000      |
|14                              |Chop-suey Chinese|374.8333      |
|80                              |Tortuga Restaurante|357.4167      |


## 6. Sales Analysis with CTEs

Assume we have the Northwind database which contains tables like Orders, OrderDetails, and Products. Create a CTE that calculates the total sales for each product in the year 1997.

In [56]:
query =  """
SELECT products.ProductName, SUM(orderdetails.productID) AS TotalSales
FROM products
JOIN orderdetails ON products.productID = orderdetails.productID
JOIN orders ON orderdetails.OrderID = orders.OrderID
WHERE YEAR(orders.orderdate) = 1997
GROUP BY 
    products.ProductName
ORDER BY 
    TotalSales DESC;
"""

df = pd.read_sql_query(query, engine)
df

ProductName  TotalSales
0             Gnocchi di nonna Alice       280.0
1               Raclette Courdavault       236.0
2                       Lakkalikööri       228.0
3          Wimmers gute Semmelknödel       192.0
4                          Tourtière       162.0
5                      Perth Pasties       159.0
6    Original Frankfurter grüne Soße       154.0
7                      Longlife Tofu       148.0
8                        Fløtemysost       142.0
9                          Spegesild       138.0
10                  Gudbrandsdalsost       138.0
11               Scottish Longbreads       136.0
12                      Vegie-spread       126.0
13                    Tarte au sucre       124.0
14                    Sirop d'érable       122.0
15                    Ravioli Angelo       114.0
16               Gumbär Gummibärchen       104.0
17                           Geitost        99.0
18                          Maxilaku        98.0
19                     Zaanse koeken        94.0
20                  Boston Crab Meat        80.0
21        Teatime Chocolate Biscuits        76.0
22                     Côte de Blaye        76.0
23              Rhönbräu Klosterbier        75.0
24                        Röd Kaviar        73.0
25            Mozzarella di Giovanni        72.0
26                     Outback Lager        70.0
27                    Steeleye Stout        70.0
28         Louisiana Hot Spiced Okra        66.0
29  Louisiana Fiery Hot Pepper Sauce        65.0
30               Sir Rodney's Scones        63.0
31                 Camembert Pierrot        60.0
32           Thüringer Rostbratwurst        58.0
33                 Rössle Sauerkraut        56.0
34                      Pâté chinois        55.0
35                      Alice Mutton        51.0
36                  Valkoinen suklaa        50.0
37                         Chocolade        48.0
38                           Pavlova        48.0
39                      Gula Malacca        44.0
40                    Queso Cabrales        44.0
41                              Tofu        42.0
42     Singaporean Hokkien Fried Mee        42.0
43   Jack's New England Clam Chowder        41.0
44                        Gravad lax        37.0
45                       Inlagd Sill        36.0
46                     Sasquatch Ale        34.0
47                 Gorgonzola Telino        31.0
48             Nord-Ost Matjeshering        30.0
49                Schoggi Schokolade        27.0
50                          Tunnbröd        23.0
51               Gustaf's Knäckebröd        22.0
52                             Konbu        13.0
53         Queso Manchego La Pastora        12.0
54                   Mishi Kobe Niku         9.0
55                             Chang         6.0
56                     Aniseed Syrup         3.0
57                             Chais         2.0

#### Expected result

|ProductName                     |TotalSales|
|--------------------------------|----------|
|Gnocchi di nonna Alice          |173       |
|Tourtière                       |126       |
|Geitost                         |119       |
|Chang                           |115       |
|Raclette Courdavault            |115       |
|Sirop d'érable                  |106       |
|Vegie-spread                    |100       |
|Côte de Blaye                   |99        |
|Alice Mutton                    |97        |
|Steeleye Stout                  |95        |
|Sir Rodney's Scones             |92        |
|Pavlova                         |86        |
|Zaanse koeken                   |85        |
|Fløtemysost                     |75        |
|Tarte au sucre                  |75        |
